### ISSUE TITLE
[Feature request] Better command line completion for aur commands

### ISSUE THREAD

**Issue by VorpalBlade** [AladW/aurutils#984]
Currently the aur bash and zsh completion appears to be auto generated, using a simplified version of the zsh _argument format as the "native" format, with a wrapper for bash.

However, the only information provided to the argument completion system is what flags exist and if they they take an argument or not. What is missing is completion of those arguments. For example `aur-query --by` can only take a limited number of different arguments. What would be nice is if that information was provided to the completion system somehow, and not just be found in the man page.

Another nice feature to have would be for the documentation of the flags to be available to the completion system, zsh can use that to display helpful descriptions of the various options. I don't believe such support is available in bash, but I could be wrong as I don't use bash. This might not be feasible unless the man pages becomes (partially) generated from the same source as the completion scripts, but that is something to consider.

To add this two things would be required:
1. Extending `--dump-options`, presumably in a reasonably low-maintenance form. Currently the implementation seems to be duplicated between the various aur scripts, which is probably not the best way to do it, especially if the implementation of it becomes more complex. Another file in /usr/share could solve this easily (I saw that aur already sources files from `/usr/share/makepkg`, so adding a `/usr/share/aurutils` seems natural).
2. Extend the bash completion script to deal with more parts of the zsh completion format.

There is an added bonus to introducing a system like this, at least for options that can take a small number of fixed different values: Any such list can also be used in the script itself for error checking to detect misspelled options. This does not seem to be done today for the most part.

And for other types of options, based on a quick scan through the man pages they seem to fall into three categories:
* Package names (possible multiple comma separated).
* File/directory names.
* "Un-completable" (job numbers, generic arguments to pass on to other commands, etc).

Thus I believe it should be possible to keep the maintenance burden of such a feature reasonably small. One problem might be the bash completion script side of things though, I lack the expertise with bash to determine how much work it would be to support such features (or possibly just ignore the unsupported parts).

I have not attached an aurutils.log, as it is not relevant to a feature request.

**Comment by rafasc** [comment]
To give a bit more context, the completion we have comes from the early 2.x. days.
At the time the aur interface was changing so quickly we opted into some crude automated solution for bash.

The zsh (which I rarely use) completion came as an afterthought.

From my limited experience writing these completions, I do not recommend going down the road you propose.
Making `--dump-options` smarter will only complicate parsing on the completion side.

I don't think the change would be hard to implement, however If we have to somehow annotate the type of  command options receive and/or centralize this feature into its own utility, I would say the burden added to maintain this "pseudo-database" would be around the same ballpark of maintaining proper completion scripts for bash and zsh.

Additionally, the zsh completion system has lots of features that could improve the completion on aur completion on zsh.

**Comment by VorpalBlade** [comment]
> I would say the burden added to maintain this "pseudo-database" would be around the same ballpark of maintaining proper completion scripts for bash and zsh.

I'm fairly familiar with zsh completion and I could definitely write manual such scripts if help is needed (I might not have the time to though, this depends). I have no knowledge about the bash completion system at all.

> Additionally, the zsh completion system has lots of features that could improve the completion on aur completion on zsh.

Agreed.

**Comment by AladW** [comment]
For me it's fine if we limit our focus on the zsh completion if more people are familiar with it. I'm not familiar with either the zsh or bash completion system I'm afraid, but I'm happy to help out with testing and feedback.

On Tue, 12 Apr 2022 05:53:01 -0700
Arvid Norlander ***@***.***> wrote:

> > I would say the burden added to maintain this "pseudo-database" would be around the same ballpark of maintaining proper completion scripts for bash and zsh.
> 
> I'm fairly familiar with zsh completion and I could definitely write manual such scripts if help is needed (I might not have the time to though, this depends). I have no knowledge about the bash completion system at all.
> 
> > Additionally, the zsh completion system has lots of features that could improve the completion on aur completion on zsh.
> 
> Agreed.
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/AladW/aurutils/issues/984#issuecomment-1096689517
> You are receiving this because you are subscribed to this thread.
> 
> Message ID: ***@***.***>


-- 
alad ***@***.***>

**Comment by VorpalBlade** [comment]
I have done a draft of what this could look like for zsh, only implementing completion for the top level and the sub command aur search. In addition there are some TODOs in it (with comments describing the issue in question). Hopefully this should allow you to evaluate if this is the path forward you want to take. It will be a larger maintenance burden than auto generated completion, and there is a risk of it getting out of sync (as happened with the man page). On the other hand, the extra features of the completion are quite nice.

I put the code up as a gist for now: https://gist.github.com/VorpalBlade/e96df739d10df88f14d799fb28a7a5b9

**Comment by AladW** [comment]
Looking at the TODOs, the main thing that would need addressing for me is the delay in `__aur_list_pkgs()`. I think a TTL of 24hours is fine. Also I would argue tab completing packages makes more sense for aur-sync, or at least should also be enabled there.

> The alternative (which I would consider if it was my code) would be to define everything about arguments in one place (including help strings, valid argument values etc) and then auto generate completion, parameter sections of man pages and argument parsing from it.

I don't know how complicated it is, but this seems something I would have to maintain and learn how it works in the future. Considering that after all this time I'm not familiar with the basic zsh completion system, I doubt that it's a good idea in practice...

**Comment by VorpalBlade** [comment]
> I don't know how complicated it is, but this seems something I would have to maintain and learn how it works in the future. Considering that after all this time I'm not familiar with the basic zsh completion system, I doubt that it's a good idea in practice...

That is a reasonable argument. My thoughts on this were basically that you would be able to auto generate both zsh and bash completion from a central source. And man pages etc. This would help keep things in sync.

And the documentation for the zsh completion system is very dense. I don't understand it on a deep level either. The whole tags system is a bit of mystery to me for example...

**Comment by AladW** [comment]
> The time for aur pkglist to run. Even when it does nothing it takes a moment. It would be faster to operate directly on ~/.cache/aurutils/pkglist/packages and only call aur pkglist if that file has an mtime that is too old.

https://github.com/AladW/aurutils/pull/987

> All that said, the pkglist tab completion is quite fast as long as it doesn't have to show too many options (i.e. when there are leading characters that narrows down the options sufficiently). All this indicates to me that most of the time is actually spent on formatting and displaying, something which the completion code has no control over.

Could the pkglist completion be enabled only when at least a single character is present?

**Comment by VorpalBlade** [comment]
> Could the pkglist completion be enabled only when at least a single character is present?

Interesting idea, I think it should be possible, I'll look into it.

> Alternatively, aur pkglist could implement something similar to type=suggest queries (auto-completion on the AUR website), in that it returns only a small set of packages for a given query. Basically aur pkglist | grep -E '^$1' | head -20.

That might also be worth looking into, possibly in combination with the above. 

I also have not looked to determine what the start up cost is of aur (sourcing scripts etc) but I did notice a difference between reading the file directly and reading it through the pkglist command. I'm not at my computer at the moment but I might do some measuring and profiling later.

**Comment by AladW** [comment]
Reading the file directly gives me:
```
$ time cat ~/.cache/aurutils/pkglist/packages >/dev/null

real	0m0.004s
user	0m0.001s
sys	0m0.004s
```
vs. with #987
```
$ time ./aur-pkglist --ttl 1800 >/dev/null 

real	0m0.013s
user	0m0.006s
sys	0m0.009s
```
and when a new package list needs retrieving:
```
$ time ./aur-pkglist --ttl 1800 >/dev/null 
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  368k  100  368k    0     0  1809k      0 --:--:-- --:--:-- --:--:-- 1814k

real	0m0.432s
user	0m0.141s
sys	0m0.059s
```

**Comment by AladW** [comment]
Merged #948 :tada: 

I'd say the bash completions are topics for another day/issue, having a much improved completion for zsh is already a great step. And I changed my shell to zsh accordingly. :+1:
